### PR TITLE
Bluetooth: hci_driver: Fix deadlock when system runs out of evt bufs

### PR DESCRIPTION
**Pull request by joerchan**
Fix deadlock between the hci_core rx_thread and the hci_driver
rx_thread.

The hci_core rx_thread is during processing of the encryption change
event will send new HCI commands to the controller and waits on a
semaphore for the command complete event.
The hci_core tx_thread successfully sends this to the controller so
it is now up to the hci_driver rx_thread to fetch the command complete
event.
The hci_driver rx_thread is waiting for an available buffer for HCI
events after having used up all the HCI event buffers from the pool
and is now waiting for the hci_core rx_thread to free up buffers.

Handle this by marking advertising reports and the QoS conn event
reports as discardable events. In a combined build this make the
hci_driver retrieve the buffers for these events for a different HCI
event buffer pool. Also do not wait for available buffers for these
events but discard if no buffer is available.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

### CODE REVIEW

**Comment by joerchan** [comment]
@MarekPieta 

If you look at the BUF count, you could probably reduce the regular buffer count and increase the discardable buffer count. I'm uncertain about how much though. Maybe @rugeGerritsen could provide some hint as to how many bufs could potentially be generated?
```
CONFIG_BT_RX_BUF_COUNT=10
CONFIG_BT_RX_BUF_LEN=76
CONFIG_BT_DISCARDABLE_BUF_COUNT=3
CONFIG_BT_DISCARDABLE_BUF_SIZE=45
```

File: subsys/bluetooth/controller/hci_driver.c [hunk 1]
```diff
@@ -192,22 +193,43 @@ static void data_packet_process(u8_t *hci_buf)
...
+			return false;
+		}
+	}
+	case BT_HCI_EVT_VENDOR:
+	{
+		u8_t subevent = hci_buf[2];
+
+		switch (subevent) {
+		case HCI_VS_SUBEVENT_CODE_QOS_CONN_EVENT_REPORT:
+			return true;
```
Comment (rugeGerritsen) [inline subsys/bluetooth/controller/hci_driver.c:218]: Should we add build asserts to ensure that the discardable buffer size is large enough? 
We could consider doing that for all events.
```suggestion
			BUILD_ASSERT(sizeof(BT_HCI_EVT_HDR_SIZE + 1 +HCI_VS_SUBEVENT_CODE_QOS_CONN_EVENT_REPORT)
			< CONFIG_BT_DISCARDABLE_BUF_SIZE);
			return true;
```

**Comment by rugeGerritsen** [comment]
> CONFIG_BT_RX_BUF_COUNT=10
> CONFIG_BT_RX_BUF_LEN=76
> CONFIG_BT_DISCARDABLE_BUF_COUNT=3
> CONFIG_BT_DISCARDABLE_BUF_SIZE=45

Using the defaults should be fine. The QoS report is raised every millisecond. Setting the discardable buffer count to 3 would give the host 3 ms to fetch the event, which I believe is sufficient

**Comment by joerchan** [comment]
> Using the defaults should be fine. The QoS report is raised every millisecond. Setting the discardable buffer count to 3 would give the host 3 ms to fetch the event, which I believe is sufficient

I meant 10 is probably too high for CONFIG_BT_RX_BUF_COUNT now. For the zephyr controller this defaults to 3.

**Comment by rugeGerritsen** [comment]
> Using the defaults should be fine. The QoS report is raised every millisecond. Setting the discardable buffer count to 3 would give the host 3 ms to fetch the event, which I believe is sufficient
> 
> I meant 10 is probably too high for CONFIG_BT_RX_BUF_COUNT now. For the zephyr controller this defaults to 3.

I agree. Using zephyr defaults should be fine

File: subsys/bluetooth/controller/hci_driver.c (hunk 1 above)
Comment (joerchan) [reply subsys/bluetooth/controller/hci_driver.c:218]: If you enable assert then net_buf_add_mem will assert if writing past the buffer.
I think that is good enough. It should be caught quickly enough.
The default for the discardable buffer size is the largest advertising report.

**Comment by joerchan** [comment]
@rugeGerritsen Zephyr controller defaults to 3, Nordic controller defaults to 10.
```
	default 3 if BT_RECV_IS_RX_THREAD
	default 10
```

**Pull request merged**